### PR TITLE
test(db): Thin line for new sql testing suite

### PR DESCRIPTION
**Pull request by tmessi**
This sets up the configuration for running a new suite of tests that are
written in sql, to test the sql logic implemented in plpgsql functions
and triggers. It is primarily designed to provides tests for the data
warehouse functionality that is implemented purely in sql.

The infrasturture to run these tests leverages docker. It creates one
docker container using the `library/postgres` image and provisions the
database with the boundary sql migrations and test helper functions.
A separate docker container is run that has the pgtap extension and
pg_prove cli to execute the tests.

As noted, pgtap is used for the tests. This postgresql extension
provides a number of helpful assertion functions to make writting sql
tests easier. It also, along with the pg_prove cli, provides readable
test output.

This commit also provides an initial test for the wh_user_dimension to
establish an inital pattern for these kind of tests.

Lastly this configures circleci to run this test suite as a separate job
in parallel with the other tests.

See:
    https://pgtap.org/

### CODE REVIEW

File: internal/db/sqltest/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# SQL Tests
```
Comment (mgaffney) [inline internal/db/sqltest/README.md:1]: Thanks for adding this README. Very helpful!

**Review by mgaffney — APPROVED** [review]
Looks great!

File: .circleci/config/jobs/test-sql.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+machine:
+  image: 'ubuntu-1604:201903-01'
```
Comment (jefferai) [inline .circleci/config/jobs/test-sql.yml:2]: Not that it necessarily matters but any reason to use such an old machine image? LTS is reasonable but why not 20.04?

File: internal/scheduler/additional_verification_test.go [hunk 3]
```diff
@@ -338,3 +339,102 @@ func TestSchedulerMonitorLoop(t *testing.T) {
 	require.NoError(err)
 	assert.Equal(string(job.Interrupted), run.Status)
 }
+
+func TestSchedulerFinalStatusUpdate(t *testing.T) {
```
Comment (jefferai) [inline internal/scheduler/additional_verification_test.go:343]: Did this come in from merging in from `main` or something? Or is this relevant to the current PR in a way I'm not seeing?

It's not a problem if it's the former, but _if so_ you may want to get in the habit of doing a `main` -> `base long-term branch` -> `feature branch` merge instead of merging directly from `main` to `feature branch` to keep PRs clean.

**Review by jefferai — COMMENTED** [review]
A couple of comments, more questions than anything else. As a non-DB expert this looks good to me!

File: internal/scheduler/additional_verification_test.go (hunk 3 above)
Comment (tmessi) [reply internal/scheduler/additional_verification_test.go:343]: Yes this is from merging `main`. I will cleanup this PR to follow that pattern of `main` -> `long-term` to clean this up.

File: .circleci/config/jobs/test-sql.yml (hunk 2 above)
Comment (tmessi) [reply .circleci/config/jobs/test-sql.yml:2]: I picked this to just follow what the [existing jobs was using](https://github.com/hashicorp/boundary/blob/tmessi-wh-test-baseline/.circleci/config/jobs/build.yml) to be consistent.

File: internal/scheduler/additional_verification_test.go (hunk 3 above)
Comment (tmessi) [reply internal/scheduler/additional_verification_test.go]: cleaned the PR up

**Review by jimlambrt — APPROVED** [review]
Ty!  Really amazing work and a good foundation.
